db/openshift.deploy.yml

```diff
@@ -326,3 +329,95 @@ objects:
               dnsPolicy: "ClusterFirst"
               serviceAccountName: "${JOB_SERVICE_ACCOUNT}"
               serviceAccount: "${JOB_SERVICE_ACCOUNT}"
+  - kind: CronJob
+    apiVersion: "batch/v1"
+    metadata:
+      name: ${NAME}-${ZONE}-${COMPONENT}-restore
+      labels:
+        app: ${NAME}-${ZONE}
+        cronjob: ${NAME}-${ZONE}
+    spec:
+      schedule: "0 0 31 2 *" # At 00:00 on day-of-month 31 in February, so it won't run, we'll trigger it manually when need
+      concurrencyPolicy: "Replace"
+      successfulJobsHistoryLimit: ${{SUCCESS_JOBS_HISTORY_LIMIT}}
+      failedJobsHistoryLimit: ${{FAILED_JOBS_HISTORY_LIMIT}}
+      jobTemplate:
+        metadata:
+          labels:
+            app: ${NAME}-${ZONE}
+            cronjob: ${NAME}-${ZONE}
+        spec:
+          backoffLimit: ${{JOB_BACKOFF_LIMIT}}
+          template:
+            metadata:
+              labels:
+                app: ${NAME}-${ZONE}
+                cronjob: ${NAME}-${ZONE}
+            spec:
+              containers:
+                - name: ${NAME}-${ZONE}-${COMPONENT}-restore
+                  # use the same image as our database, so we can run the psql command
+                  image: image-registry.apps.silver.devops.gov.bc.ca/${OC_NAMESPACE}/${NAME}-${ZONE}-${COMPONENT}:${ZONE}-db
+                  command: ["/bin/sh", "-c"]
+                  args:
+                  - |
+                    sql_file=$(find backups/${BACKUP_FILE_DIR} -type f -name "*.sql" -print0)
+                    if [  -z $sql_file ]; then
+                      sql_zip_file=$(find backups/${BACKUP_FILE_DIR} -type f -name "*.sql.gz" -print0)
+                      if [ -z $sql_zip_file ]; then
+                        echo "No zipped backup file found"
+                      else
+                        echo "Found zipped backup file: $sql_zip_file"
+                        gzip -dk $sql_zip_file
+                      fi
+                    fi
+                    sql_file=$(find backups/${BACKUP_FILE_DIR} -type f -name "*.sql" -print0)
+                    if [  -z $sql_file ]; then
+                      echo "No backup SQL file is found"
+                    else
+                      echo "Found SQL file, rename existing database, and create a new empty database"
+                      psql -h ${NAME}-${ZONE}-${COMPONENT} -U ${POSTGRES_USER} -c "ALTER DATABASE fom RENAME TO ${RENAME_DATABASE_TO};" -c "CREATE DATABASE fom;"
+                      echo "Running SQL file: $sql_file"
+                      psql -h ${NAME}-${ZONE}-${COMPONENT} -U ${POSTGRES_USER} -d ${POSTGRES_DB} -f $sql_file
+                      echo "Finish database restore"
+                    fi
```

Comment:
So not the logic is that:

1. Try to find the sql file first (in case already unzip it in the past)
2. If not found, try to find the zipped sql file, and unzip it
3. Get the sql file again, if not found then print not found
4. If found then run restore script

Reply:
Should exit with error if no backup SQL file is found. Should check the return value of the psql commands and report errors.

Reply:
The restore finishes with the old schema existing, which is fine. Part of the recovery process could be to remove that. Also, maybe this process should perform a backup first before doing the restore?

Reply:
when no backup file is found, it will just print "Not found" and do nothing for now. I'll check how to report errors.
<img width="819" alt="Screen Shot 2024-08-20 at 2 41 06 PM" src="https://github.com/user-attachments/assets/04bcb165-d58a-46e1-85b9-c01852993ff0">

Reply:
So rather than having the cron job finish successfully in the case of no backup SQL file found, we should have the job run marked as a failure (which I assume will happen with a non-zero exit code from the script).

Reply:
When no backup file found, exit 1:
<img width="815" alt="Screen Shot 2024-08-21 at 4 05 42 PM" src="https://github.com/user-attachments/assets/ec11bc40-b75d-4eff-aeba-6194c943cc11">

When psql command failed, exit 1 (the following example failure happened when I pass an empty file `psql -h fom-24-db -U ${POSTGRES_USER} -d ${POSTGRES_DB} -f ''`):

<img width="963" alt="Screen Shot 2024-08-21 at 4 12 54 PM" src="https://github.com/user-attachments/assets/e4b6da58-763a-4ae1-8443-e88c9b1531a0">

Reply:
I think it's better to run backup before restore, but not sure if we want to include that in this script as well, it's getting complicated....

Reply:
Yes, probably sufficient for now to just have our documented process say to do a manual backup before trying to do the restore. (and bring the app down before doing the backup).